### PR TITLE
Add “Authorization server” in the side menu of the BO

### LINKED ISSUE #29809: APIS07 - Add “Authorization server” in the side menu of the BO
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

This is a new feature. It aims to make the authorization server management accessible in the BO

#### Description

As a shop administrator
I want to access the configuration of my authorization server
In order to manage my API accesses

#### Business rules
N/A

#### Acceptance test

**AT1**
1. Given that 
    1. I am a shop administrator
    2. The external authorization server administration module is installed
2. When I go in the BO
3. Then
    1. I can access the external authz server administration page through the left side menu bar
    2. I can access the external authz server administration page through the configuration of the module in the module management page

**AT2**
1. Given that 
    1. I am a shop administrator
    2. The external authorization server administration module isn’t installed
2. When I go in the BO
3. Then
    1. the module access doesn’t appear in the left side menu bar
    2. the native authz server administration page is accessible

**AT1**
1. Given that 
    1. I am a shop administrator
    2. The external authorization server administration module is installed
    3. The external authorization server administration module is deactivated
2. When I go in the BO
3. Then 
    1. the module access doesn’t appear in the left side menu bar
    2. the native authz server administration page is accessible

### Alternatives you've considered

_No response_

### Additional context

_No response_

### Do you plan to work on this feature?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

### PR DESCRIPTION
**Pull request by mflasquin**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add “Authorization server” in the side menu of the BO
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #29809 
| Related PRs       | -
| How to test?      | See #29809 
| Possible impacts? | -

### CODE REVIEW

File: src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/AuthorizationServer/index.html.twig [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ *#}
+{% extends '@PrestaShop/Admin/layout.html.twig' %}
+
+{% block javascripts %}
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/AuthorizationServer/index.html.twig]: useless block

**Review by FabienPapet — COMMENTED** [review]
One minor fix and it's all good on my side

File: classes/lang/KeysReference/FeatureFlagLang.php [hunk 2]
```diff
@@ -42,3 +42,7 @@
 // Legacy link feature flags
 trans('Attribute groups', 'Admin.Advparameters.Feature');
 trans('Enable or disable the migrated attribute groups page.', 'Admin.Advparameters.Help');
+
+// Authorization server feature flag
+trans('Authorization server', 'Admin.Advparameters.Feature');
+trans('Enable or disable the authorization server page.', 'Admin.Advparameters.Help');
```
Comment (l-delin) [inline classes/lang/KeysReference/FeatureFlagLang.php:48]: Hello, **"Enable or disable the authorization server page."** Is this a toggle button? If yes, is its default position "off"? Thanks :-)

Comment (mflasquin) [reply classes/lang/KeysReference/FeatureFlagLang.php:48]: Hi, yes off by default :)

**Comment by mflasquin** [comment]
I would like to get your attention to this commit that allows you to benefit from creating subscriber events that can be connected to prestashop hooks : https://github.com/PrestaShop/PrestaShop/pull/30337/commits/d5b17ed843c472ed47ad137f142542b9e55fb568

File: classes/lang/KeysReference/FeatureFlagLang.php (hunk 2 above)
Comment (l-delin) [inline classes/lang/KeysReference/FeatureFlagLang.php:48]:
```suggestion
trans('Enable authorization server page.', 'Admin.Advparameters.Help');
```

File: install-dev/data/xml/feature_flag.xml [hunk 3]
```diff
@@ -13,5 +13,6 @@
...
     <feature_flag id="product_page_v2_multi_shop" name="product_page_v2_multi_shop" label_wording="New product page - Multi store" label_domain="Admin.Advparameters.Feature" description_wording="Access the new product page, even in a multistore context. This is a work in progress and some features are not available." description_domain="Admin.Advparameters.Help" state="0" stability="beta" />
     <feature_flag id="attribute_group" name="attribute_group" label_wording="Attribute group" label_domain="Admin.Advparameters.Feature" description_wording="Enable / Disable migrated attribute group page." description_domain="Admin.Advparameters.Help" state="0" stability="beta" />
+    <feature_flag id="authorization_server" name="authorization_server" label_wording="Authorization server" label_domain="Admin.Advparameters.Feature" description_wording="Enable or disable the authorization server page." description_domain="Admin.Advparameters.Help" state="0" stability="beta" />
```
Comment (l-delin) [inline install-dev/data/xml/feature_flag.xml:16]:
```suggestion
    <feature_flag id="authorization_server" name="authorization_server" label_wording="Authorization server" label_domain="Admin.Advparameters.Feature" description_wording="Enable authorization server page." description_domain="Admin.Advparameters.Help" state="0" stability="beta" />
```

**Comment by nicosomb** [comment]
@mflasquin Could you rebase your branch please, with CI fixes?

File: install-dev/data/xml/tab.xml [hunk 4]
```diff
@@ -609,6 +609,11 @@
                     <wording>Webservice</wording>
                     <wording_domain>Admin.Navigation.Menu</wording_domain>
                 </tab>
+                <tab id="AuthorizationServer" id_parent="Advanced_Parameters" active="0" enabled="1">
```
Comment (atomiix) [inline install-dev/data/xml/tab.xml:612]: Maybe hide it by default?
```suggestion
                <tab id="AuthorizationServer" id_parent="Advanced_Parameters" active="0" enabled="0">
```

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AuthorizationServerController.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Manages the "Configure > Advanced Parameters > Authorization Server" page.
+ */
+class AuthorizationServerController extends FrameworkBundleAdminController
+{
+    /**
+     * @return Response
+     */
```
Comment (atomiix) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AuthorizationServerController.php:44]: There's probably an `@AdminSecurity` annotation to add here

File: src/PrestaShopBundle/Resources/config/services/bundle/services.yml [hunk 6]
```diff
@@ -61,9 +61,7 @@ services:
     class: PrestaShop\PrestaShop\Adapter\Hook\HookDispatcher
     arguments:
       - '@request_stack'
-    calls:
-      - [ addSubscriber, [ "@prestashop.adapter.legacy.hook.subscriber" ] ]
-      - [ addSubscriber, [ "@prestashop.adapter.legacy.block.helper.subscriber" ] ]
+      - !tagged core.legacy.hook.subscriber
```
Comment (atomiix) [inline src/PrestaShopBundle/Resources/config/services/bundle/services.yml:64]: Nice! :)

File: install-dev/data/xml/tab.xml (hunk 4 above)
Comment (mflasquin) [reply install-dev/data/xml/tab.xml:612]: It hide with the "active" param :)

File: install-dev/data/xml/feature_flag.xml (hunk 3 above)
Comment (mflasquin) [reply install-dev/data/xml/feature_flag.xml:16]: The button allows you to deactivate or activate the functionality and not only activate it

File: classes/lang/KeysReference/FeatureFlagLang.php (hunk 2 above)
Comment (mflasquin) [reply classes/lang/KeysReference/FeatureFlagLang.php:48]: The button allows you to deactivate or activate the functionality and not only activate it

File: classes/Tab.php [hunk 7]
```diff
@@ -304,7 +304,7 @@ public static function getModuleTabList()
      *
      * @return array tabs
      */
-    public static $_cache_tabs = [];
+    protected static $_cache_tabs = [];
```
Comment (FabienPapet) [inline classes/Tab.php]: This is a BC

File: src/Adapter/Hook/HookDispatcher.php [hunk 8]
```diff
@@ -59,10 +60,17 @@ class HookDispatcher extends EventDispatcher implements HookDispatcherInterface
...
      * @param RequestStack|null $requestStack (nullable to preserve backward compatibility)
+     * @param iterable|null $hookSubscribers
      */
-    public function __construct(RequestStack $requestStack = null)
+    public function __construct(RequestStack $requestStack = null, iterable $hookSubscribers = null)
     {
         $this->requestStack = $requestStack;
+
+        foreach ($hookSubscribers as $hookSubscriber) {
+            if ($hookSubscriber instanceof EventSubscriberInterface) {
```
Comment (FabienPapet) [inline src/Adapter/Hook/HookDispatcher.php]: useless check imo

File: src/PrestaShopBundle/Resources/config/services/bundle/services.yml (hunk 6 above)
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/config/services/bundle/services.yml:64]: can you add a `prestashop.` prefix on your tag name ?

File: src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/FeatureFlag/EventListener/FeatureFlagTypeListener.php [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    public function onFeatureFlagBetaSave(HookEvent $event): void
+    {
+        if (isset($event->getHookParameters()['form_data']['feature_flags']['authorization_server']['enabled'])) {
+            $this->commandBus->handle(
+                new UpdateTabStatusByClassNameCommand(
+                    'AdminAuthorizationServer',
+                    $event->getHookParameters()['form_data']['feature_flags']['authorization_server']['enabled']
+                )
+            );
+            TabDataProvider::resetTabCache();
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/FeatureFlag/EventListener/FeatureFlagTypeListener.php]: Static call here. Maybe use DI

File: src/Adapter/Tab/TabDataProvider.php [hunk 10]
```diff
@@ -131,4 +131,14 @@ private function canAccessTab($profileId, $tabId, $accessLevel = 'view')
 
         return false;
     }
+
+    /**
+     * Reset static tab cache
+     *
+     * @return void
+     */
+    public static function resetTabCache(): void
```
Comment (FabienPapet) [inline src/Adapter/Tab/TabDataProvider.php]: this does not need to be static

File: classes/Tab.php (hunk 7 above)
Comment (mflasquin) [reply classes/Tab.php]: It was already protected before changes. It just an error in my commit, finally if you check the changes, there is no change on this property

File: src/PrestaShopBundle/Resources/config/services/bundle/services.yml (hunk 6 above)
Comment (mflasquin) [reply src/PrestaShopBundle/Resources/config/services/bundle/services.yml:64]: We have another tags beginning with "core", I think we have to keep a coherence ? ex : core.product_command_builder

File: src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/FeatureFlag/EventListener/FeatureFlagTypeListener.php (hunk 9 above)
Comment (mflasquin) [reply src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/FeatureFlag/EventListener/FeatureFlagTypeListener.php]: Can you develop your idea a little more?

File: src/PrestaShopBundle/Resources/config/services/bundle/services.yml (hunk 6 above)
Comment (FabienPapet) [reply src/PrestaShopBundle/Resources/config/services/bundle/services.yml:64]: I think it was a mistake to not prefix them :/

File: src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/FeatureFlag/EventListener/FeatureFlagTypeListener.php (hunk 9 above)
Comment (mflasquin) [reply src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/FeatureFlag/EventListener/FeatureFlagTypeListener.php]: done :smile:

File: src/Adapter/Tab/TabDataProvider.php (hunk 10 above)
Comment (mflasquin) [reply src/Adapter/Tab/TabDataProvider.php]: done :smile:

File: src/Adapter/Hook/HookDispatcher.php (hunk 8 above)
Comment (mflasquin) [reply src/Adapter/Hook/HookDispatcher.php]: done :smile:

**Review by FabienPapet — DISMISSED** [review]

File: src/Adapter/Tab/TabDataProvider.php [hunk 11]
```diff
@@ -35,23 +37,23 @@
...
+    private $legacyContext;
 
     /**
-     * @param int $contextEmployeeProfileId
-     * @param int $superAdminProfileId
+     * @param LegacyContext $legacyContext
+     * @param Configuration $legacyConfiguration
      */
-    public function __construct($contextEmployeeProfileId, $superAdminProfileId)
+    public function __construct(LegacyContext $legacyContext, Configuration $legacyConfiguration)
```
Comment (FabienPapet) [inline src/Adapter/Tab/TabDataProvider.php]: Can you use the ConfigurationInterface class (in core NS) ?

**Review by matks — APPROVED** [review]

**Review by FabienPapet — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Comment by AureRita** [comment]
Hi @mflasquin 

Thank you for this PR, I tested it and I see some little things as you can see on this video : 

[Untitled_ Jan 3, 2023 3_07 PM.webm](https://user-images.githubusercontent.com/100699445/210373468-4067b40b-2760-437e-8b9e-dd0c03ae903e.webm)


- You can see that the button help didn't open on a little tab, it'll replace the current webpage 
- You can see that the name of the section isn't translate even if you put a languages already here 

Did this issue should be corrected on this PR on in another one after ? 

Waiting for your feedback

**Comment by mflasquin** [comment]
> Hi @mflasquin
> 
> Thank you for this PR, I tested it and I see some little things as you can see on this video :
> 
>  Untitled_.Jan.3.2023.3_07.PM.webm 
> * You can see that the button help didn't open on a little tab, it'll replace the current webpage
> * You can see that the name of the section isn't translate even if you put a languages already here
> 
> Did this issue should be corrected on this PR on in another one after ?
> 
> Waiting for your feedback

Hi @AureRita, the help button is fixed in https://github.com/PrestaShop/PrestaShop/pull/30497 the PR dedicated for this page
Also we don't add translations in PRs

**Review by AureRita — APPROVED** [review]
hi @mflasquin 

Thank for this information, so I can put it in Qa Approved :heavy_check_mark: 

Thank you

**Comment by mflasquin** [comment]
Thanks @AureRita for the QA :smile:

**Pull request merged**